### ISSUE TITLE
Updating cache causes "Error: socket hang up"

### ISSUE THREAD

**Issue by IvanFon** [tldr-pages/tldr-node-client#152]
## Expected behaviour
The cache should be updated.

## Actual behaviour
A socket error, shown below.

## Log, debug output
```
❯ tldr -V
2.0.1

❯ tldr --update
Updating...
Error: socket hang up
    at createHangUpError (_http_client.js:345:15)
    at Socket.socketOnEnd (_http_client.js:437:23)
    at emitNone (events.js:110:20)
    at Socket.emit (events.js:207:7)
    at endReadableNT (_stream_readable.js:1059:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)

❯ tldr tar
Page not found. Updating cache ..
Error: socket hang up
    at createHangUpError (_http_client.js:345:15)
    at Socket.socketOnEnd (_http_client.js:437:23)
    at emitNone (events.js:110:20)
    at Socket.emit (events.js:207:7)
    at endReadableNT (_stream_readable.js:1059:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

## Environment
- Arch Linux
- Node.js v8.6.0

**Comment by agnivade** [comment]
Can you tell me the repro steps for this ?

I am unable to repro this. 

```
09:25:24-agniva-~$sudo npm -g install tldr
[sudo] password for agniva: 
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
/usr/local/bin/tldr -> /usr/local/lib/node_modules/tldr/bin/tldr
- jodid25519@1.0.2 node_modules/tldr/node_modules/jodid25519
/usr/local/lib
`-- tldr@2.0.1 
  +-- fs-extra@0.30.0
  | `-- rimraf@2.6.2 
  |   `-- glob@7.1.2 
  |     `-- minimatch@3.0.4 
  |       `-- brace-expansion@1.1.8 
  |         `-- balanced-match@1.0.0 
  `-- request@2.75.0
    +-- extend@3.0.1 
    +-- har-validator@2.0.6
    | `-- is-my-json-valid@2.16.1 
    +-- http-signature@1.1.1
    | +-- jsprim@1.4.1 
    | | +-- extsprintf@1.3.0 
    | | `-- verror@1.10.0 
    | |   `-- assert-plus@1.0.0 
    | `-- sshpk@1.13.1 
    |   `-- getpass@0.1.7 
    +-- mime-types@2.1.17 
    | `-- mime-db@1.30.0 
    `-- tough-cookie@2.3.3 

09:26:17-agniva-~$tldr --update
Updating...
Done
09:26:25-agniva-~$tldr tar

  tar

  Archiving utility.
  Often combined with a compression method, such as gzip or bzip.

  - Create an archive from files:
    tar cf target.tar file1 file2 file3

  - Create a gzipped archive:
    tar czf target.tar.gz file1 file2 file3

  - Extract an archive in a target folder:
    tar xf source.tar -C folder

  - Extract a gzipped archive in the current directory:
    tar xzf source.tar.gz

  - Extract a bzipped archive in the current directory:
    tar xjf source.tar.bz2

  - Create a compressed archive, using archive suffix to determine the compression program:
    tar caf target.tar.xz file1 file2 file3

  - List the contents of a tar file:
    tar tvf source.tar

```

**Comment by IvanFon** [comment]
It seems I'm doing the same thing:

```
❯ sudo npm -g install tldr
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
/usr/bin/tldr -> /usr/lib/node_modules/tldr/bin/tldr
+ tldr@2.0.1
added 120 packages in 1.528s

❯ tldr -V
2.0.1

❯ tldr --update
Updating...
Error: socket hang up
    at createHangUpError (_http_client.js:345:15)
    at Socket.socketOnEnd (_http_client.js:437:23)
    at emitNone (events.js:110:20)
    at Socket.emit (events.js:207:7)
    at endReadableNT (_stream_readable.js:1059:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

Could this be a Node issue? Are there any ways I can test this further?

**Comment by agnivade** [comment]
I just did a fresh install of node on a new machine and retried your steps.
```
ubuntu@ip-10-0-0-80:~⟫ sudo tar -C /usr/local --strip-components 1 -xJf node-v8.6.0-linux-x64.tar.xz
ubuntu@ip-10-0-0-80:~⟫ node -v
v8.6.0
ubuntu@ip-10-0-0-80:~⟫ sudo npm install -g tldr
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
/usr/local/bin/tldr -> /usr/local/lib/node_modules/tldr/bin/tldr
+ tldr@2.0.1
added 120 packages in 5.233s
ubuntu@ip-10-0-0-80:~⟫ tldr tar
Page not found. Updating cache ..

  tar

  Archiving utility.
  Often combined with a compression method, such as gzip or bzip.

  - Create an archive from files:
    tar cf target.tar file1 file2 file3

  - Create a gzipped archive:
    tar czf target.tar.gz file1 file2 file3

  - Extract an archive in a target folder:
    tar xf source.tar -C folder

  - Extract a gzipped archive in the current directory:
    tar xzf source.tar.gz

  - Extract a bzipped archive in the current directory:
    tar xjf source.tar.bz2

  - Create a compressed archive, using archive suffix to determine the compression program:
    tar caf target.tar.xz file1 file2 file3

  - List the contents of a tar file:
    tar tvf source.tar


ubuntu@ip-10-0-0-80:~⟫ tldr --update
Updating...
Done
```

It seems like a node issue. But I'm not exactly sure what is it. Because the stack trace is not even coming from the client's code. Can you manually try to download this file and see whether it works `http://tldr-pages.github.io/assets/tldr.zip` ?

You can also download a file from the main tldr repo and test whether it can atleast render a file locally. `tldr --render /path/to/file.md`.

**Comment by IvanFon** [comment]
Sorry for the late response.

I cloned `tldr-pages/tldr`, but rendering pages from there doesn't seem to work:

```
~/.tldr
❯ ls pages/common | grep tar.md
tar.md

~/.tldr
❯ tldr --render pages/common/tar.md 
/usr/lib/node_modules/tldr/lib/tldr.js:137
  if (options.markdown) {
             ^

TypeError: Cannot read property 'markdown' of undefined
    at renderContent (/usr/lib/node_modules/tldr/lib/tldr.js:137:14)
    at index.getShortIndex (/usr/lib/node_modules/tldr/lib/tldr.js:66:7)
    at ReadFileContext.fs.readFile (/usr/lib/node_modules/tldr/lib/index.js:110:16)
    at ReadFileContext.callback (/usr/lib/node_modules/tldr/node_modules/graceful-fs/graceful-fs.js:78:16)
    at FSReqWrap.readFileAfterOpen [as oncomplete] (fs.js:421:13)
```

Where should I put the `tldr.zip` file? From reading the code, it seems like it should be unzipped and go in `~/.tldr`, but that didn't change anything.

This doesn't really seem to be a tldr issue.

**Comment by agnivade** [comment]
Oops I think you caught a bug. I really didnt test the `--render` option for a long time. Let me push a patch.

No need to put the zip file anywhere,  I just wanted you to check whether you can download the file at all in the first place. Since you were having some socket hangup issues, I wanted to confirm whether its not a network related issue.

**Comment by agnivade** [comment]
Hey @IvanFon - I just published 2.0.2. Would you mind re-installing and trying the `--render` option again ?

**Comment by IvanFon** [comment]
`--render` is working great now!

On the updating cache issue, I'll close this for now since the problem doesn't seem reproducible and the stack trace doesn't originate from tldr. If I find out anything else I'll comment here later. Thanks for the help :)
